nada_numpy/types.py

```diff
@@ -1415,3 +2571,503 @@ def get_log_scale() -> int:
 def reset_log_scale() -> None:
     """Resets the Rational log scaling factor to the original default value"""
     _RationalConfig.log_scale = _RationalConfig.default_log_scale
+
+
+# Fixed-point math operations
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+#
+# Part of the code is from the CrypTen Facebook Project:
+# https://github.com/facebookresearch/CrypTen/blob/main/crypten/common/functions/logic.py
+# https://github.com/facebookresearch/CrypTen/blob/main/crypten/common/functions/approximations.py
+#
+# Modifications:
+# July, 2024
+#   - Nada datatypes.
+#   - Relative accuracy documentation.
+#   - Some performance improvements.
+#   - Fixed Tanh Chebyshev method by changing '_hardtanh' implementation.
+#   - Tan.
+#   - Motzkin's prolynomial preprocessing approach.
+#   - GeLU and SiLU functions.
+
+
+def sign(x: _NadaRational) -> _NadaRational:
+    """Computes the sign value (0 is considered positive)"""
+
+    ltz_cond = x < rational(0)
+    ltz = ltz_cond.if_else(rational(1), rational(0))
+
+    return rational(1) - rational(2) * ltz
```

Comment:
These multiplications seem to cause rescaling? Or is it avoided somehow?

Reply:
It causes rescaling yes, and it should cause that. But we can avoida that in fact by doing 1 - ltz - ltz in fact. I will update that

Reply:
It causes rescaling and it should cause rescaling. But we can avoid truncation by doing 1 - ltz - ltz in fact. I'll update that

Reply:
Indeed as @nilmemo mentions, this does rescale after every multiplication. I am not sure if the specific math of polynomials can be improved into a somewhat more efficient use of rotations, but, in `nada-numpy` we have this thing called `UnsafeArithmeticSession` which creates an environment where auto-rescaling does not occur if there's a better way to do it. You can find more details on the implementation in `nada_numpy/context.py`. We use it for matrix multiplication for `Rational` types as every multiplication triggers unnecessary rescaling. In case you want to use it, this is the way:

```python
A = rational(4.0)
B = rational(5.0)
with UnsafeArithmeticSession():
    result_not_rescaled = A * B # within this context A * B does not trigger rescaling
    result = result_not_rescaled.rescale_down()
```